hail/hail/src/is/hail/io/index/StagedIndexReader.scala

```diff
@@ -645,3 +646,26 @@ class StagedIndexReader(
     ).asBaseStruct
   }
 }
+
+class SpeccedStagedIndexReader(
+  emb: EmitMethodBuilder[_],
+  spec: AbstractIndexSpec,
+) extends StagedIndexReader(emb, spec.leafCodec, spec.internalNodeCodec) {
```

Comment:
From a very quick look, it appears like it might not be hard to just replace `StagedIndexReader` with the specced version. I think it would be preferable both to let `StagedIndexReader` be a final class, and to keep the leaf and node codecs bundled together. What do you think?

Reply:
We'd need to create the appropriate index spec for bgen readers, which currently don't have one.

Reply:
Ah yeah, I see. Up to you if you want to try that. I'm happy to approve if you send it back to me.

Reply:
I did it. Wasn't hard.